lib/kernel/src/gen_tcp.erl

```diff
@@ -218,21 +223,89 @@ connect(#{family := Fam} = SockAddr, Opts, Timeout)
       Reason  :: timeout | inet:posix().
 
 connect(Address, Port, Opts0, Timeout) ->
-    case inet:gen_tcp_module(Opts0) of
+    %% When neither `inet` nor `inet6` is provided in Opts0,
+    %% and if `ipv6_probe` option is given, try to connect ipv6 first.
+    {TryIpv6, Ipv6T} =
+        case proplists:get_value(ipv6_probe, Opts0) of
+                  true -> {true, 2000}; %% default 2 seconds
+                  false -> {false, 0};
+                  undefined -> {false, 0};
+                  T -> {true, T}
+              end,
+    %% delete it to avoid interference
+    Opts1 = proplists:delete(ipv6_probe, Opts0),
+    case inet:gen_tcp_module(Opts1) of
         {?MODULE, Opts} ->
-            Timer = inet:start_timer(Timeout),
-            Res = (catch connect1(Address,Port,Opts,Timer)),
-            _ = inet:stop_timer(Timer),
-            case Res of
-                {ok,S} -> {ok,S};
-                {error, einval} -> exit(badarg);
-                {'EXIT',Reason} -> exit(Reason);
-                Error -> Error
-            end;
+            connect_maybe_ipv6(Address, Port, Opts, Timeout, TryIpv6, Ipv6T);
         {GenTcpMod, Opts} ->
             GenTcpMod:connect(Address, Port, Opts, Timeout)
     end.
 
+connect_maybe_ipv6(Address, Port, Opts, Timeout, TryIpv6, Ipv6T) ->
+    case maybe_ipv6(Address, Opts, TryIpv6) of
+        {maybe, NewOpts} when TryIpv6 ->
+            try
+                {ok, _} = connect_0(Address, Port, NewOpts, Ipv6T)
+            catch
+                _ : _ ->
+                    %% fallback
+                    connect_0(Address, Port, Opts, Timeout)
+            end;
+        NewOpts ->
+            connect_0(Address, Port, NewOpts, Timeout)
+    end.
+
+connect_0(Address, Port, Opts, Timeout) ->
+    Timer = inet:start_timer(Timeout),
+    Res = (catch connect1(Address,Port,Opts,Timer)),
+    _ = inet:stop_timer(Timer),
+    case Res of
+        {ok,S} -> {ok,S};
+        {error, einval} -> exit(badarg);
+        {'EXIT',Reason} -> exit(Reason);
+        Error -> Error
+    end.
+
+maybe_ipv6({local, _}, Opts, _TryIpv6) ->
+    %% unapplicable to local sockets
+    Opts;
+maybe_ipv6(Host, Opts, TryIpv6) ->
+    case lists:member(inet, Opts) orelse lists:member(inet6, Opts) of
+        true ->
+            Opts; %% caller has made the decision
+        false when is_tuple(Host) ->
+            %% ip tuple provided
+            maybe_ipv6_1(Host, Opts);
+        false when TryIpv6 ->
+            %% string host
+            maybe_ipv6_2(Host, Opts);
+        false ->
+            Opts
+    end.
+
+maybe_ipv6_1(Ip, Opts) when tuple_size(Ip) =:= 4 -> Opts;
+maybe_ipv6_1(Ip, Opts) when tuple_size(Ip) =:= 8 -> [inet6 | Opts].
```

Comment:
Please help to fix this `function_clause` error when a bad size tuple is provided.
Something happens in wolff tests.
```suggestion
maybe_ipv6_1(Ip, Opts) when tuple_size(Ip) =:= 8 -> [inet6 | Opts];
maybe_ipv6_1(_, Opts) -> Opts;
```

Reply:
```
                   {function_clause,
                       [{gen_tcp,maybe_ipv6_1,
                            [{127,0,0},[{active,false},binary]],
                            [{file,"gen_tcp.erl"},{line,278}]},
                        {gen_tcp,connect_maybe_ipv6,6,
                            [{file,"gen_tcp.erl"},{line,237}]},
                        {kpro_connection,connect,4,
                            [{file,
                                 "/mnt/code/src/kafka/wolff/_build/default/lib/kafka_protocol/src/kpro_connection.erl"},
                             {line,224}]},
                        {kpro_connection,init,4,
                            [{file,
                                 "/mnt/code/src/kafka/wolff/_build/default/lib/kafka_protocol/src/kpro_connection.erl"},
                             {line,195}]},
                        {proc_lib,init_p_do_apply,3,
                            [{file,"proc_lib.erl"},{line,240}]}]}},
```

Reply:
nvm.
the original code crashes with a function_clause anyway:

```
 {function_clause,[{inet,mod,
                                           [infinity,tcp_module,
                                            {undefined,{127,0,0,1}},
                                            #{local => local_tcp,inet => inet_tcp,inet6 => inet6_tcp},
                                            undefined,[]],
                                           [{file,"inet.erl"},{line,1478}]},
                                     {gen_tcp,connect1,4,[{file,"gen_tcp.erl"},{line,237}]},
                                     {gen_tcp,connect,4,[{file,"gen_tcp.erl"},{line,224}]},
                                     {erl_eval,do_apply,7,[{file,"erl_eval.erl"},{line,750}]},
                                     {shell,exprs,7,[{file,"shell.erl"},{line,782}]},
                                     {shell,eval_exprs,7,[{file,"shell.erl"},{line,738}]},
                                     {shell,eval_loop,4,[{file,"shell.erl"},{line,723}]}]}

```